Firmware/depth-sensing-JSN-SR04_Gen3_RK-Integration-Indu

```diff
@@ -0,0 +1,326 @@
+#include <Adafruit_MAX31865.h>
+const int TEMP_CHIP_SELECT = D6;
+Adafruit_MAX31865 sensor = Adafruit_MAX31865(TEMP_CHIP_SELECT);
+#define RREF 430.0 //Rref resistor value = 430.0
+
+
+//-------------------------- LED Setup -------------------------------------------
+const pin_t MY_LED = D7; // blink to let us know you're alive
+bool led_state = HIGH; // starting state
+
+
+// Global objects; TODO: save power stats
+FuelGauge batteryMonitor;
+
+SYSTEM_THREAD(ENABLED);
+SYSTEM_MODE(SEMI_AUTOMATIC);
+
+// Declare one of these as a global variable to manage the JSN-SR04 sensor
+JSN_SR04_Gen3 distanceSensor;
+
+
+//===============================================================================
+// INITIALIZATION
+//===============================================================================
+void setup() {
+
+    Serial.begin(115200);
+
+    // Initialize the sensor configuration from setup()
+    distanceSensor
+        .withTrigPin(TRIG_PIN)
+        .withEchoPin(ECHO_PIN)
+        .withUnusedPins(UNUSED_PIN0, UNUSED_PIN3)
+        .withCallback(distanceCallback)
+        .withSamplePeriodic(500ms)
+        .setup();
+
+        // Set up GPS
+        GPS.begin(9600);
+
+        // uncomment next line to turn on RMC (recommended minimum) and GGA (fix data) including altitude
+        GPS.sendCommand(PMTK_SET_NMEA_OUTPUT_RMCGGA);
+
+        // uncomment next line to turn on only the "minimum recommended" data
+        //GPS.sendCommand(PMTK_SET_NMEA_OUTPUT_RMCONLY);
+
+        // Set the update rate
+        GPS.sendCommand(PMTK_SET_NMEA_UPDATE_1HZ); // 1 Hz update rate
+        // For the parsing code to work nicely and have time to sort thru the data, and print it out we don't suggest using anything higher than 1 Hz
+    
+        // Initialize the SD library
+        if (!SD.begin(SD_CHIP_SELECT, SPI_FULL_SPEED)) {
+            Serial.println("failed to open card");
+            return;
+        }
+
+}
+
+
+//===============================================================================
+// MAIN
+//===============================================================================
+void loop() {
+    // You must call this frequently from loop(), preferable on every execution
+    distanceSensor.loop();
+   
+    // read data from the GPS in the 'main loop'
+    char c = GPS.read();
+
+    // if a sentence is received, we can check the checksum, parse it...
+    if (GPS.newNMEAreceived()) {
+        if (!GPS.parse(GPS.lastNMEA())) // this also sets the newNMEAreceived() flag to false
+            return; // we can fail to parse a sentence in which case we should just wait for another
+    }
+
+    // approximately every second or so, print out the current stats
+    if (millis() - timer > 1000) {
+        timer = millis(); // reset the timer
+
+        serialPrintGPSTime();
+
+        temperatureC = getTemp();
+        JSN_SR04_Gen3::DistanceResult result = distanceSesnor.getDistance();
+        depth = distanceCallback(result);
```

Comment:
Here, in the main loop, I have called distanceCallback() function to get the depth in cm. Later, in line 129, when printToFile() is called, the depth will be saved and written onto the SD card.